api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,275 @@
+syntax = "proto3";
+
+package envoy.service.mutation.v3alpha;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto";
+import "envoy/service/auth/v3/attribute_context.proto";
+import "envoy/type/v3/http_status.proto";
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/struct.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.service.mutation.v3alpha";
+option java_outer_classname = "MutationProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+option (udpa.annotations.file_status).work_in_progress = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: External Mutation Service]
+
+// A service that can mutate HTTP requests and responses.
+// The overall mutation protocol works like this:
+//
+// 1. Envoy sends to the service information about the HTTP request.
+// 2. The service sends back a MutationResponse message that directs Envoy
+//    to either stop processing, continue without it, or send it the
+//    next chunk of the message body.
+// 3. If so requested, Envoy sends the server a chunk of the message body.
+//    and the server sends back a MutationResponse.
+// 4. If so requested, Envoy sends the server the HTTP trailers,
+//    and the server sends back a MutationResponse.
+// 5. At this point, request processing is done, and we pick up again
+//    at step 1 when Envoy receives a response from the upstream server.
+// 6. At any point above, if the server closes the gRPC stream cleanly,
+//    then Envoy proceeds without consulting the server.
+// 7. At any point above, if the server closes the gRPC stream with an error,
+//    then Envoy returns a 500 error to the client, unless the filter
+//    was configured to ignore errors.
+//
+// In other words, the process is a request/response conversation, but
+// using a gRPC stream to make it easier for the client and server to
+// maintain state. The exception is the "SEND_BODY_ASYNC" command,
+// which requests that the server keep sending body chunks without waiting
+// for a response.
+
+service Mutation {
+  // This begins the bidirectional stream that Envoy will use to
+  // give the server control over what the filter does. The actual
+  // protocol is described by the MutationRequest and MutationResponse
+  // messages below.
+  rpc Mutate(stream MutationRequest) returns (stream MutationResponse) {
+  }
+}
+
+// This represents the different types of messages that Envoy can send
+// to a mutation server.
+// [#next-free-field: 8]
+message MutationRequest {
+  oneof r {
```

Comment:
`oneof` seems like an unnecessary limitation. Since this is sent to an external service, which might be a few ms away, it could be useful e.g. to send a complete request (i.e. request headers + complete request body + request trailers), which isn't possible with `oneof`.